### PR TITLE
Fix: [ATen] add more missing moves - part 2

### PR DESCRIPTION
**Pull request by Skylion007**
Applies some more missing std::move found by static analysis. This should improve performance and reduce unnecessary copies. This PR only targets ATen for now.

And before you ask about the edits, std::move is optimal in a ternary operator as copy ellision cannot happen one. The best thing is probably rewriting it as an if else, but ultimately this should be performant enough.
Followup to #88512 and #88514 

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel @VitalyFedyunin @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/FunctionalTensorWrapper.cpp [hunk 1]
```diff
@@ -529,7 +531,7 @@ bool isFunctionalTensorIListRef(c10::IListRef<T> list) {
 }
 
 bool isFunctionalTensor(ITensorListRef list) {
-  return isFunctionalTensorIListRef(list);
+  return isFunctionalTensorIListRef(std::move(list));
```
Comment (swolchok) [inline aten/src/ATen/FunctionalTensorWrapper.cpp:534]: I don't understand how IListRef is movable at all currently, but this looks like the wrong fix -- isFunctionalTensorIListRef is not trying to take ownership

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -3355,27 +3356,27 @@ linalg_svd_out(const Tensor& A,
...
   // The only way to achieve these two things and still abide by the compositionality rules
   // is by dispatching to another function.
-  return at::_linalg_svd_out(U, S, Vh, A, full_matrices, /*compute_uv=*/true, driver);
+  return at::_linalg_svd_out(U, S, Vh, A, full_matrices, /*compute_uv=*/true, std::move(driver));
 }
 
 std::tuple<Tensor, Tensor, Tensor> linalg_svd(const Tensor& A, bool full_matrices,
     c10::optional<c10::string_view> driver) {
-  return at::_linalg_svd(A, full_matrices, /*compute_uv=*/true, driver);
+  return at::_linalg_svd(A, full_matrices, /*compute_uv=*/true, std::move(driver));
```
Comment (swolchok) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3364]: I don't think moving optional<string_view> does anything

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -3355,27 +3356,27 @@ linalg_svd_out(const Tensor& A,
...
 }
 
 // See note in linalg_svd for why this function does not have an _ex variant
 Tensor& linalg_svdvals_out(const Tensor& A, c10::optional<c10::string_view> driver, Tensor & S) {
   // Dummies
   auto U = at::empty({0}, A.options());
   auto Vh = at::empty({0}, A.options());
-  at::_linalg_svd_out(U, S, Vh, A, /*full_matrices=*/false, /*comptue_uv=*/false, /*driver=*/driver);
+  at::_linalg_svd_out(U, S, Vh, A, /*full_matrices=*/false, /*comptue_uv=*/false, /*driver=*/std::move(driver));
```
Comment (swolchok) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3372]: ditto

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 4]
```diff
@@ -3355,27 +3356,27 @@ linalg_svd_out(const Tensor& A,
...
   return S;
 }
 
 Tensor linalg_svdvals(const Tensor& A, c10::optional<c10::string_view> driver) {
   return std::get<1>(at::_linalg_svd(A, /*full_matrices=*/false,
                      /*compute_uv=*/_may_require_fw_or_bw_grad(A),
-                     /*driver=*/driver));
+                     /*driver=*/std::move(driver)));
```
Comment (swolchok) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3379]: ditto

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 5]
```diff
@@ -3715,7 +3716,7 @@ std::tuple<Tensor&, Tensor&, Tensor&, Tensor&> linalg_lstsq_out(
   // 'singular_values' is expected to have real float dtype
   checkLinalgCompatibleDtype("torch.linalg.lstsq", singular_values.scalar_type(), real_dtype, "singular_values");
 
-  std::string driver_name = get_default_lstsq_driver(driver, input);
+  std::string driver_name = get_default_lstsq_driver(std::move(driver), input);
```
Comment (swolchok) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3719]: ditto

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 6]
```diff
@@ -3838,8 +3839,8 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> linalg_lstsq(
   Tensor rank = at::empty({0}, input.options().dtype(at::kLong));
   Tensor singular_values = at::empty({0}, input.options().dtype(toRealValueType(input.scalar_type())));
   std::tie(solution, residuals, rank, singular_values) =
-      at::linalg_lstsq_outf(input, other, rcond, driver, solution, residuals, rank, singular_values);
-  return std::make_tuple(solution, residuals, rank, singular_values);
+      at::linalg_lstsq_outf(input, other, rcond, std::move(driver), solution, residuals, rank, singular_values);
```
Comment (swolchok) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3842]: ditto

File: aten/src/ATen/native/SpectralOps.cpp [hunk 7]
```diff
@@ -205,7 +206,7 @@ Tensor fft_c2r(c10::string_view function_name,
   if (n_opt) {
     input = resize_fft_input(input, dim, n/2 + 1);
   }
-  const auto norm = norm_from_string(norm_str, forward);
+  const auto norm = norm_from_string(std::move(norm_str), forward);
```
Comment (swolchok) [inline aten/src/ATen/native/SpectralOps.cpp:209]: lots more move of optional<string_view> in this file, please fix all

**Review by swolchok — CHANGES_REQUESTED** [review]
mostly good. i like that you caught a bunch of missing moves of SymInt in particular; CC @ezyang on those since they're easy to introduce. changes requested for unnecessary moves pointed out in comments

**Comment by ezyang** [comment]
Thanks! Deferring to swolchok here

File: aten/src/ATen/FunctionalTensorWrapper.cpp (hunk 1 above)
Comment (Skylion007) [reply aten/src/ATen/FunctionalTensorWrapper.cpp]: Actually, taking const ref causes overload ambiguity interestingly enough.

**Review by ezyang — APPROVED** [review]

**Comment by Skylion007** [comment]
Hmm, @ezyang saw a failure in Torch Dynamo (that only occurs an hour after CI starts oof). Any way to just run the tests that failed?

**Comment by Skylion007** [comment]
https://github.com/pytorch/pytorch/actions/runs/3659725602/jobs/6186425064

**Comment by ezyang** [comment]
in GHA ui you can ask to rerun failed tests

**Comment by Skylion007** [comment]
> in GHA ui you can ask to rerun failed tests

I meant for new commits that attempt to fix it, or do you believe it's just a flake?

**Comment by ezyang** [comment]
oh sorry, we're just gonna run everything, not your problem lol.

I don't know if it's your problem or not, this PR is so big it's hard to tell. Maybe we should split it in two or something.

**Comment by Skylion007** [comment]
Ah, it wasn't my fault. Let me revert that change quickly.: https://github.com/pytorch/pytorch/commit/b2795d3c4eeff929a368f7b51492d0cfab2c3547

**Comment by Skylion007** [comment]
@ezyang Reverted and merged fixes form master, just kick off the merge process again. ;) Should have realized it wasn't my fault due to it being from the trunk jobs. ;)

**Comment by Skylion007** [comment]
Okay, @ezyang merge conflicts dealt with and the only failing test so far is the a flake due to an download error during install.

**Comment by Skylion007** [comment]
> in GHA ui you can ask to rerun failed tests

BTW, it seems I sadly do not have permissions to. :/

**Pull request closed without merging**